pkg/util/external_storage.go

```diff
@@ -90,3 +96,81 @@ func DefaultS3Retryer() request.Retryer {
 		},
 	}
 }
+
+// RemoveFilesIf removes files from external storage if the path matches the predicate.
+func RemoveFilesIf(
+	ctx context.Context,
+	extStorage storage.ExternalStorage,
+	pred func(path string) bool,
+	opt *storage.WalkOption,
+) error {
+	var toRemoveFiles []string
+	err := extStorage.WalkDir(ctx, opt, func(path string, _ int64) error {
+		path = strings.TrimPrefix(path, "/")
+		if pred(path) {
+			toRemoveFiles = append(toRemoveFiles, path)
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.ErrExternalStorageAPI.Wrap(err).GenWithStackByArgs("RemoveTemporaryFiles")
+	}
+
+	log.Debug("Removing files", zap.Any("toRemoveFiles", toRemoveFiles))
+
+	for _, path := range toRemoveFiles {
+		if err := extStorage.DeleteFile(ctx, path); err != nil {
+			return errors.ErrExternalStorageAPI.Wrap(err)
+		}
+	}
+	return DeleteFilesInExtStorage(ctx, extStorage, toRemoveFiles)
+}
+
+// DeleteFilesInExtStorage deletes files in external storage concurrently.
+func DeleteFilesInExtStorage(
+	ctx context.Context, extStorage storage.ExternalStorage, toRemoveFiles []string,
+) error {
+	eg, eCtx := errgroup.WithContext(ctx)
+	for _, file := range toRemoveFiles {
+		name := file
+		// TODO: limit number of goroutines
```

Comment:
Please review this:
Perhaps [a new concurrency control package](https://github.com/sourcegraph/conc) could be introduced here to simplify the code?